timeless/views.py

```diff
@@ -30,86 +49,148 @@ def delete(self):
         pass
 
 
-class GenericView(views.View):
+class GenericView(views.MethodView):
     """ Generic view with common logic """
     template_name = None
     methods = ["get", "post"]
-    permissions = ()
 
-    def dispatch_request(self, *args, **kwargs):
-        """This method is called with all the arguments from the URL rule."""
-        http_method = request.method.lower()
-        if http_method not in self.methods:
-            raise Exception("Method is not allowed")
-
-        if not self.check_permissions():
-            raise Exception("Permissions weren't passed")
-
-        method = getattr(self, http_method, None)
-        if not method:
-            raise Exception(f"Provide {http_method} method")
+    @classmethod
+    def register(cls, blueprint, route, name=None):
+        """
+        A shortcut method for registering this view to an app or blueprint.
+        Assuming we have a blueprint and a CompanyCreate view, then these two
+        lines are identical in functionality:
+            views.add_url_rule('/companies/create',
+                               view_func=CompanyCreate.as_view(
+                                   'company_create')
+                               )
+            CompanyCreate.register(views, '/companies/create',
+                                   'company_create')
+        """
+        if not name:
+            # Convert "ViewName" to "view_name" and use it
+            name = camel_to_underscore.sub(r"_\1", cls.__name__).lower()
+            blueprint.add_url_rule(route, view_func=cls.as_view(name))
+
+    def dispatch(self):
+        """
+        Hook for a subclass to call before dispatch actually happens.
+        """
+        pass
 
-        return method(*args, **kwargs)
+    def dispatch_request(self, *args, **kwargs):
+        """
+        Save args and kwargs, then dispatch the request as a normal MethodView,
+        calling get() or post().
+        """
+        self.args = args
+        self.kwargs = kwargs
 
-    def check_permissions(self):
-        """ Method for permissions checking """
-        for permission in self.permissions:
```

Comment:
@bkuzmic btw I think checking permission here in `dispatch_request` is a good way. And using it something like this:
```python
class FloorListView(ListView):
    model = Floor
    tempalte_name = '/templates/floors/list.html'
    permission_classes = (IsAuthenticated,)
```
wdyt?

Reply:
@vryazanov I thought that it would be better to implement this using already provided decorators, like for this example:

class AdminView(View):
    decorators = [auth.admin_required, auth.login_required]

class UserView(AdminView):
    ....

WDYT? Maybe I can leave a puzzle for that.

Reply:
@bkuzmic ah, I didnt know about decorators. It would be great if you leave puzzle :+1: